### PR TITLE
Don't use raw parameter types in `find_builder_fn`

### LINKED ISSUE #121314: ICE: `param types with same index differ in name`
auto-reduced (treereduce-rust):
````rust
use std::collections::VecDeque;

pub fn yes_vecdeque_partial_eq_ref_mut_array<'a, A, B>() -> impl PartialEq {
    VecDeque::<A>::q_user()
}


````

original:
````rust
//@ check-pass

pub fn yes_vec_partial_eq_array<A, B>() -> impl PartialEq<[B; 32]>
where
    A: PartialEq<B>,
{
    Vec::<A>::new()
}

pub fn yes_vec_partial_eq_ref_array<'a, A, B>() -> impl PartialEq<&'a [B; 32]>
where
        Inner2: Default,
{
    Vec::<A>::new()
}

pub fn yes_array_into_vec<T>() -> Vec<T> {
    [].into()
}

pub fn yes_array_into_box<T>() -> Box<[T]> {
    [].into()
}

use std::collections::VecDeque;

pub fn yes_vecdeque_partial_eq_array<A, B>() -> impl PartialEq<[B; 32]>
where
    A: PartialEq<B>,
{
    VecDeque::<A>::new()
}

pub fn yes_vecdeque_partial_eq_ref_array<'a, A, B>() -> impl PartialEq<&'a [B; 32]>
where
    A: PartialEq<B>,
{
    VecDeque::<A>::new()
}

pub fn yes_vecdeque_partial_eq_ref_mut_array<'a, A, B>() -> impl PartialEq<&'writes_to_path mut [B; tokenize]>
where
    A: PartialEq<B>,
{
    VecDeque::<A>::q_user()
}

fn main() {}

````

Version information
````
rustc 1.78.0-dev
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.78.0-dev
LLVM version: 18.1.0

````

Command:
`/home/matthias/.rustup/toolchains/local-debug-assertions/bin/rustc `

### PR DESCRIPTION
**Pull request by compiler-errors**
We shouldn't really ever be using `EarlyBinder::skip_binder` then performing type equality, since param types will never be equal to other types. When checking compatibility with the signature, we instead create some fresh args.

Fixes #121314

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 1]
```diff
@@ -1597,7 +1597,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             .filter(|item| matches!(item.kind, ty::AssocKind::Fn) && !item.fn_has_self_parameter)
             .filter_map(|item| {
                 // Only assoc fns that return `Self`, `Option<Self>` or `Result<Self, _>`.
-                let ret_ty = self.tcx.fn_sig(item.def_id).skip_binder().output();
+                let ret_ty = self
+                    .tcx
+                    .fn_sig(item.def_id)
+                    .instantiate(self.tcx, self.fresh_args_for_item(DUMMY_SP, item.def_id))
+                    .output();
```
Comment (oli-obk) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:1604]:
```suggestion
                    .output()
                    .instantiate(self.tcx, self.fresh_args_for_item(DUMMY_SP, item.def_id));
```

**Review by oli-obk — APPROVED** [review]

File: compiler/rustc_hir_typeck/src/method/suggest.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:1604]: EarlyBinder<FnSig>::output() doesnt exist

Comment (oli-obk) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:1604]: Oh heh. I was looking at polyfnsig

Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:1604]: yeah I think youre thinking of the late binder, not early binder haha

**Pull request merged**